core/src/main/java/it/pagopa/selfcare/onboarding/core/InstitutionServiceImpl.java

```diff
@@ -118,13 +118,25 @@ public void onboardingPaAggregator(OnboardingData onboardingData) {
     }
 
     @Override
-    public void onboardingCompanyV2(OnboardingData onboardingData) {
+    public void onboardingCompanyV2(OnboardingData onboardingData, String userFiscalCode) {
         log.trace("onboardingProductAsync start");
         log.debug("onboardingProductAsync onboardingData = {}", onboardingData);
+        InstitutionInfoIC userBusinesses = partyRegistryProxyConnector.getInstitutionsByUserFiscalCode(userFiscalCode);
+        if(businessIsNotRelatedToUser(userBusinesses, onboardingData.getTaxCode())){
+            throw new OnboardingNotAllowedException("The selected business does not belong to the user");
+        }
         onboardingMsConnector.onboardingCompany(onboardingData);
         log.trace("onboarding end");
     }
 
+    private boolean businessIsNotRelatedToUser(InstitutionInfoIC institutionInfoIC, String businessTaxCode) {
+        return institutionInfoIC == null
+                || CollectionUtils.isEmpty(institutionInfoIC.getBusinesses())
+                || institutionInfoIC.getBusinesses()
+                .stream()
+                .noneMatch(business -> business.getBusinessTaxId().equals(businessTaxCode));
```

Comment:
If either `businessTaxId` is null, this can lead to a NullPointerException. you could use Objects.equals() to safely compare strings, which handles null values gracefully.

`.noneMatch(business -> Objects.equals(business.getBusinessTaxId(), businessTaxCode));`